remove unused Base.GMP.MPZ.import! and test Base.GMP.MPZ.export!

NegaScout opened the pull request:
rename Base.GMP.MPZ.export! "dynamic" variable from `a` to `x`, since it is more in line with comment on L141 in base/gmp.jl
https://github.com/JuliaLang/julia/blob/cab11bbd2d5452f523ef277bfee2affec4e3c9f5/base/gmp.jl#L141-L146

add tests, that numbers and vectors can be exported and imported correctly and composed import and export is identity

LilithHafner commented:
Thanks!

AFAICT, `import!` is an internal function that is neither used nor tested. Perhaps it should be removed? As of 6d71254, this PR makes a breaking change to `import!` so if we run pkgeval we should be able to tell if anyone depends on this internal function. `export!` is used once and untested so adding tests for it is very nice.

This invokes pkgeval: @nanosoldier `runtests()`

NegaScout commented:
Oh, I wanted to run some kind of grep at least, if it is used somewhere, but I forgot. Sorry. However, this functionality is very handy for "interpreting memory as number". I am not sure if there is more use cases. However it would be nice for any serious RSA implementation (and ElGamal at least I think, possibly more cryptographic schemes), which needs to interpret memory as a number to sign it or encrypt/decrypt it.

I would like to see movement in the opposite direction and would like to see more support for this. Otherwise, some iteration based construction for this use case would be needed, unless some Julia native BigInt type is present in Base.

I also think, any big number serialization/deserialization would benefit from this.

giordano commented:
No uses outside of Base according to https://juliahub.com/ui/Search?type=symbols&q=import!

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56903#issuecomment-2561376281) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/6d71254_vs_64d37f5/report.html) is available.

LilithHafner commented:
Generally, I lean toward avoiding new public API unless there is a very compelling case. The fact that nobody has used this internal function all this time makes me doubt that it is valuable commonly enough to warrant including in the public API. For those reasons, I prefer deleting `insert!` rather than publicizing it.

That said, it would be nice to hear other folks' opinions on whether we should keep/publicize/remove `Base.GMP.MPZ.import!`.

LilithHafner commented:
Triage talked about this and considered weather we want to provide a user facing wrapper of MPZ or if we just want to keep these internal and use only what we need to implement `BigInt`s and other Base functions. Making a good user-facing wrapper of MPZ would be fairly large and we don't want to put that in Base. Instead, it makes more sense for a fully featured MPZ wrapper to exist as a package (open invitation to anyone who wants to create that package :) ). Making `import!` public without also making many more functions public wouldn't really make sense.

For `import!`, specifically, it makes sense to delete the function from Base because we're not using it and don't want to add it to the public API.

NegaScout commented:
Ok, fair. It makes sense since there is still much in MPZ that would be cool to have in Julia but does not have bindings. So, lets rebase it and keep only the test?

NegaScout commented:
Not sure, if this is how you meant to move the definition from base to tests. However I could not define `Base_GMP_MPZ_import` as `Base.GMP.MPZ.import!` in the test, not sure why :thinking:

LilithHafner commented on base/gmp.jl:
```diff
@@ -251,21 +251,21 @@ get_str!(x, a, b::BigInt) = (ccall((:__gmpz_get_str,libgmp), Ptr{Cchar}, (Ptr{Cc
     ndigits = cld(sizeinbase(n, 2), 8*sizeof(T) - nails)
-    length(a) < ndigits && resize!(a, ndigits)
+    length(x) < ndigits && resize!(x, ndigits)
     count = Ref{Csize_t}()
-    ccall((:__gmpz_export, libgmp), Ptr{T}, (Ptr{T}, Ref{Csize_t}, Cint, Csize_t, Cint, Csize_t, mpz_t),
-        a, count, order, sizeof(T), endian, nails, n)
-    @assert count[] ≤ length(a)
-    return a, Int(count[])
+    ccall((:__gmpz_export, libgmp),
+           Ptr{T},
+           (Ptr{T}, Ref{Csize_t}, Cint, Csize_t, Cint, Csize_t, mpz_t),
+           x, count, order, sizeof(T), endian, nails, n)
+    @assert count[] ≤ length(x)
+    return x, Int(count[])
 end
```
We try to avoid committing formatting-only changes unless they are meaningful improvements and/or the old format was clearly bad. In this case, renaming this vector from `a` to `x` and changing line breaks just doesn't seem worth doing.

OTOH, I really don't care that much about this formatting so if you want to leave these changes in that's okay with me.

NegaScout replied:
I ended up removing it. It was either remove, or do separate commit and at that point it did not make sense anymore I think

LilithHafner commented on test/gmp.jl:
```diff
@@ -445,6 +445,44 @@ end
     @test string(big(0), base = rand(2:62), pad = 0) == ""
 end
 
+@testset "Base.GMP.MPZ.import!/export!" begin
```
```suggestion
@testset "Base.GMP.MPZ.export!" begin
```
At this point, the testset is only testing `export!` because `import!` isn't a thing anymore (outside of the testset itself)

LilithHafner reviewed (approved):
> Not sure, if this is how you meant to move the definition from base to tests. 

Yes, that's it exactly. It would be bad to define `Base.GMP.MPZ.import!` in tests because then someone could write tests for that function that appear to be testing a function defined in Base when that function actually does not exist in Base julia!

NegaScout committed:
remove refactor Base.GMP.MPZ.import! from base

add tests, that	numbers and vectors can be exported and imported correctly and composed import and export is identity

rfourquet commented:
For the record, `import!` was initially used in #9122 to implement `rand(::UnitRange{BigInt})` which then got simplified in #13815 by bypassing `import!` and using internals directly.

Base.GMP was never meant to be public, and `GMP.MPZ` even less so, but I'd have leaned towards keeping `MPZ.export!` as the wrapping work was already done.

LilithHafner commented:
We did keep `MPZ.export!`, and added tests for it :)

rfourquet commented:
Haha yeah, sorry I meant `import!` ;-) but great PR anyway!
